companion/src/fieldswitched.rs

```diff
@@ -0,0 +1,378 @@
+use std::fs::File;
+use std::io::{self, BufReader};
+
+mod instruction;
+mod witness;
+
+use reverie::algebra::gf2::*;
+use reverie::algebra::z64::*;
+use reverie::evaluate_fieldswitching_btoa_program;
+use reverie::Instruction;
+use reverie::{fieldswitching, ConnectionInstruction, ProgramTriple};
+
+use async_std::task;
+
+use std::process::exit;
+use std::sync::Arc;
+
+use clap::{App, Arg};
+use std::marker::PhantomData;
+use std::mem;
+
+const MAX_VEC_SIZE: usize = 1024 * 1024 * 1024;
+
+pub trait Parser<E>: Sized {
+    fn new(reader: BufReader<File>) -> io::Result<Self>;
+
+    fn next(&mut self) -> io::Result<Option<E>>;
+}
+
+fn write_vec<W: io::Write>(dst: &mut W, src: &[u8]) -> io::Result<()> {
+    assert!(src.len() < MAX_VEC_SIZE);
+    dst.write_all(&(src.len() as u32).to_le_bytes()[..])?;
+    dst.write_all(src)
+}
+
+fn read_vec<R: io::Read>(src: &mut R) -> io::Result<Option<Vec<u8>>> {
```

Comment:
`read_vec` and `write_vec` are duplicates of the ones present in `main.rs`. We should eventually extract them to a shared module, but this PR is big enough already.